File: NetDeck/DeckImport.m
//
//  DeckImport.m
//  Net Deck
//
//  Created by Gereon Steffens on 01.02.14.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

@import AFNetworking;
@import SDCAlertView;

#import "EXTScope.h"
#import "DeckImport.h"
#import "OctgnImport.h"
#import "GZip.h"

#define IMPORT_ALWAYS   NO  // set to yes for easier debugging

#if !defined(DEBUG)
#if IMPORT_ALWAYS
#warning resetting IMPORT_ALWAYS
#undef IMPORT_ALWAYS
#define IMPORT_ALWAYS   NO
#endif
#endif

typedef NS_ENUM(NSInteger, DeckBuilderSource)
{
    DeckBuilderSourceNone,
    DeckBuilderSourceNRDBList,
    DeckBuilderSourceNRDBShared,
    DeckBuilderSourceMeteor
};

@interface DeckSource : NSObject
@property NSString* deckId;
@property DeckBuilderSource source;
@end
@implementation DeckSource
@end

@interface DeckImport()

@property SDCAlertView* alert;
@property AFHTTPRequestOperationManager* manager;
@property BOOL downloadStopped;
@property DeckSource* deckSource;
@property Deck* deck;

@end

@implementation DeckImport

static DeckImport* instance;

+(DeckImport*) sharedInstance
{
    if (instance == nil)
    {
        instance = [DeckImport new];
    }
    return instance;
}

+(void) updateCount
{
    NSInteger c = [UIPasteboard generalPasteboard].changeCount;
    
    [[NSUserDefaults standardUserDefaults] setInteger:c forKey:SettingsKeys.CLIP_CHANGE_COUNT];
}

+(void) checkClipboardForDeck
{
    DeckImport* di = [DeckImport sharedInstance];
    [di checkClipboardForDeck];
}

-(void) checkClipboardForDeck
{
    UIPasteboard *pasteboard = [UIPasteboard generalPasteboard];
    
    NSInteger lastChange = [[NSUserDefaults standardUserDefaults] integerForKey:SettingsKeys.CLIP_CHANGE_COUNT];
    if (lastChange == pasteboard.changeCount && !IMPORT_ALWAYS)
    {
        return;
    }
    [[NSUserDefaults standardUserDefaults] setInteger:pasteboard.changeCount forKey:SettingsKeys.CLIP_CHANGE_COUNT];
    
    NSString* clip = pasteboard.string;

    if (clip.length == 0)
    {
        return;
    }
    
    NSArray* lines = [clip componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];
    
    self.deck = nil;
    
    self.deckSource = [self checkForNetrunnerDbDeckURL:lines];    
    if (!self.deckSource)
    {
        self.deckSource = [self checkForMeteorDeckURL:lines];
    }
    
    UIAlertController* alert = nil;
    if (self.deckSource)
    {
        if (self.deckSource.source == DeckBuilderSourceNRDBList || self.deckSource.source == DeckBuilderSourceNRDBShared)
        {
            alert = [UIAlertController alertControllerWithTitle:nil
                                                        message:l10n(@"Detected a NetrunnerDB.com deck list URL in your clipboard. Download and import this deck?")
                                                 preferredStyle:UIAlertControllerStyleAlert];
        }
        else if (self.deckSource.source == DeckBuilderSourceMeteor)
        {
            alert = [UIAlertController alertControllerWithTitle:nil
                                                        message:l10n(@"Detected a meteor deck list URL in your clipboard. Download and import this deck?")
                                                 preferredStyle:UIAlertControllerStyleAlert];
        }
    }
    else
    {
        self.deck = [self checkForTextDeck:lines];
        
        if (self.deck != nil)
        {
            alert = [UIAlertController alertControllerWithTitle:nil
                                                        message:l10n(@"Detected a deck list in your clipboard. Import this deck?")
                                                 preferredStyle:UIAlertControllerStyleAlert];
        }
    }
    
    if (!alert) {
        return;
    }
    
    
    [alert addAction:[UIAlertAction actionWithTitle:l10n(@"No") style:UIAlertActionStyleCancel handler:nil]];
    [alert addAction:[UIAlertAction actionWithTitle:l10n(@"Yes") handler:^(UIAlertAction * _Nonnull action) {
        if (self.deck)
        {
            [[NSNotificationCenter defaultCenter] postNotificationName:Notifications.IMPORT_DECK object:self userInfo:@{ @"deck": self.deck }];
        }
        else if (self.deckSource)
        {
            [self downloadDeck:self.deckSource];
        }
        self.deck = nil;
        self.deckSource = nil;

    }]];
    [alert show];
}

-(DeckSource*) checkForNetrunnerDbDeckURL:(NSArray<NSString*>*) lines
{
    // a netrunnerdb.com decklist url looks like this:
    // http://netrunnerdb.com/en/decklist/3124/in-a-red-dress-and-alone-jamieson-s-store-champ-deck-#
    // or like this:
    // http://netrunnerdb.com/en/deck/view/456867
   
    NSRegularExpression* list = [NSRegularExpression regularExpressionWithPattern:@"http://netrunnerdb.com/../decklist/(\\d*)/.*" options:0 error:nil];
    NSRegularExpression* shared = [NSRegularExpression regularExpressionWithPattern:@"http://netrunnerdb.com/../deck/view/(\\d*)" options:0 error:nil];

    NSDictionary* dict = @{
        @(DeckBuilderSourceNRDBShared): shared,
        @(DeckBuilderSourceNRDBList): list
    };
    
    
    for (NSNumber*n in dict)
    {
        DeckBuilderSource source = n.integerValue;
        NSRegularExpression* regEx = dict[n];
        
        for (NSString* line in lines)
        {
            NSTextCheckingResult* match = [regEx firstMatchInString:line options:0 range:NSMakeRange(0, [line length])];
            if (match.numberOfRanges == 2)
            {
                DeckSource* src = [[DeckSource alloc] init];
                src.deckId = [line substringWithRange:[match rangeAtIndex:1]];
                src.source = source;
                return src;
            }
        }
    }

    return nil;
}

-(DeckSource*) checkForMeteorDeckURL:(NSArray<NSString*>*) lines
{
    // a netrunner.meteor.com decklist url looks like this:
    // http://netrunner.meteor.com/decks/yBMJ3GL6FPozt9nkQ/
    // or like this (no slash)
    // http://netrunner.meteor.com/decks/i6sLkn5cYZ3633WAu
    
    NSRegularExpression* urlRegex = [NSRegularExpression regularExpressionWithPattern:@"http://netrunner.meteor.com/decks/(.*)/?" options:0 error:nil];
    
    for (NSString* line in lines)
    {
        NSTextCheckingResult* match = [urlRegex firstMatchInString:line options:0 range:NSMakeRange(0, [line length])];
        if (match.numberOfRanges == 2)
        {
            DeckSource* src = [[DeckSource alloc] init];
            src.deckId = [line substringWithRange:[match rangeAtIndex:1]];
            src.source = DeckBuilderSourceMeteor;
            return src;
        }
    }
    
    return nil;
}

-(Deck*) checkForTextDeck:(NSArray<NSString*>*)lines
{
    NSArray* cards = [CardManager allCards];
    NSRegularExpression *regex1 = [NSRegularExpression regularExpressionWithPattern:@"^([0-9])x" options:0 error:nil]; // start with "1x ..."
    NSRegularExpression *regex2 = [NSRegularExpression regularExpressionWithPattern:@" x([0-9])" options:0 error:nil]; // end with "... x3"
    NSRegularExpression *regex3 = [NSRegularExpression regularExpressionWithPattern:@"^([0-9]) " options:0 error:nil]; // start with "1 ..."
    
    NSString* name;
    
    Deck* deck = [[Deck alloc] init];
    NRRole role = NRRoleNone;
    for (NSString* line in lines)
    {
        if (name == nil)
        {
            name = line;
        }
        
        for (Card* c in cards)
        {
            // don't bother checking cards of the opposite role (as soon as we know this deck's role)
            BOOL roleOk = role == NRRoleNone || role == c.role;
            if (!roleOk) {
                continue;
            }

            NSUInteger loc = [line rangeOfString:c.name_en options:NSCaseInsensitiveSearch|NSDiacriticInsensitiveSearch].location;
            if (loc == NSNotFound)
            {
                loc = [line rangeOfString:c.name options:NSCaseInsensitiveSearch|NSDiacriticInsensitiveSearch].location;
            }
            
            if (loc != NSNotFound)
            {
                
                if (c.type == NRCardTypeIdentity)
                {
                    [deck addCard:c copies:1];
                    role = c.role;
                    // NSLog(@"found identity %@", c.name);
                }
                else
                {
                    NSTextCheckingResult *match = [regex1 firstMatchInString:line options:0 range:NSMakeRange(0, [line length])];
                    if (!match)
                    {
                        match = [regex2 firstMatchInString:line options:0 range:NSMakeRange(0, [line length])];
                    }
                    if (!match)
                    {
                        match = [regex3 firstMatchInString:line options:0 range:NSMakeRange(0, [line length])];
                    }
                    
                    if (match.numberOfRanges == 2)
                    {
                        NSString* count = [line substringWithRange:[match rangeAtIndex:1]];
                        // NSLog(@"found card %@ x %@", count, c.name);
                        
                        int cnt = [count intValue];
                        int max = deck.isDraft ? 100 : 4;
                        if (cnt > 0 && cnt < max)
                        {
                            [deck addCard:c copies:cnt];
                        }
                        
                        break;
                    }
                }
            }
        }
    }
    
    if (deck.identity != nil && deck.cards.count > 0)
    {
        deck.name = name;
        
        return deck;
    }
    else
    {
        return nil;
    }
}

#pragma mark deck data download

-(void) downloadDeck:(DeckSource*) deckSource
{
    UIView* view = [[UIView alloc] initWithFrame:CGRectMake(0,0, SDCAlertViewWidth, 20)];
    UIActivityIndicatorView* act = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
    act.center = CGPointMake(SDCAlertViewWidth/2, view.frame.size.height/2);
    [act startAnimating];
    [view addSubview:act];
    
    self.alert = [SDCAlertView alertWithTitle:l10n(@"Downloading Deck") message:nil subview:view buttons:@[l10n(@"Stop")]];
    
    @weakify(self);
    self.alert.didDismissHandler = ^(NSInteger buttonIndex) {
        @strongify(self);
        [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
        
        self.downloadStopped = YES;
        self.alert = nil;
        
        [self.manager.operationQueue cancelAllOperations];
        return;
    };

    [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
    
    if (deckSource.source == DeckBuilderSourceNRDBList) {
        [self performSelector:@selector(doDownloadDeckFromNetrunnerDbList:) withObject:deckSource.deckId afterDelay:0.01];
    }
    else if (deckSource.source == DeckBuilderSourceNRDBShared) {
        [self performSelector:@selector(doDownloadDeckFromNetrunnerDbShared:) withObject:deckSource.deckId afterDelay:0.01];
    }
    else {
        [self performSelector:@selector(doDownloadDeckFromMeteor:) withObject:deckSource.deckId afterDelay:0.01];
    }
}

-(void) doDownloadDeckFromNetrunnerDbList:(NSString*)deckId {
    NSString* deckUrl = [NSString stringWithFormat:@"http://netrunnerdb.com/api/decklist/%@", deckId];
    [self doDownloadDeckFromNetrunnerDb:deckUrl];
}

-(void) doDownloadDeckFromNetrunnerDbShared:(NSString*)deckId {
    NSString* deckUrl = [NSString stringWithFormat:@"http://netrunnerdb.com/api/shareddeck/%@", deckId];
    [self doDownloadDeckFromNetrunnerDb:deckUrl];
}

-(void) doDownloadDeckFromNetrunnerDb:(NSString*)deckUrl
{
    BOOL __block ok = NO;
    self.downloadStopped = NO;
    
    self.manager = [AFHTTPRequestOperationManager manager];
    self.manager.responseSerializer = [AFJSONResponseSerializer serializer];
    
    @weakify(self);
    [self.manager GET:deckUrl parameters:nil
              success:^(AFHTTPRequestOperation* operation, id responseObject) {
                  @strongify(self);
                  if (!self.downloadStopped)
                  {
                      // NSLog(@"deck successfully downloaded");
                      ok = [self parseJsonDecklist:responseObject];
                  }
                  [self downloadFinished:ok];
              }
              failure:^(AFHTTPRequestOperation* operation, NSError* error) {
                  @strongify(self);
                  // NSLog(@"download failed %@", operation);
                  [self downloadFinished:NO];
              }
     ];
}

-(void) doDownloadDeckFromMeteor:(NSString*)deckId
{
    NSString* deckUrl = [NSString stringWithFormat:@"http://netrunner.meteor.com/deckexport/octgn/%@/", deckId];
    BOOL __block ok = NO;
    self.downloadStopped = NO;
    
    self.manager = [AFHTTPRequestOperationManager manager];
    self.manager.responseSerializer = [AFXMLParserResponseSerializer serializer];
    NSMutableSet* contentTypes = [NSMutableSet setWithSet:self.manager.responseSerializer.acceptableContentTypes];
    [contentTypes addObject:@"application/force-download"];
    self.manager.responseSerializer.acceptableContentTypes = contentTypes;
    
    @weakify(self);
    [self.manager GET:deckUrl parameters:nil
              success:^(AFHTTPRequestOperation* operation, id responseObject) {
                  @strongify(self);
                  if (!self.downloadStopped)
                  {
                      // NSLog(@"deck successfully downloaded");
                      
                      // filename comes in Content-Disposition header, which looks like
                      // Content-Disposition: attachment; filename=Copy%20of%20Weyland%20Speed.o8d
                      NSString* disposition = operation.response.allHeaderFields[@"Content-Disposition"];
                      NSString* filename;
                      NSRange range = [disposition rangeOfString:@"filename=" options:NSCaseInsensitiveSearch];
                      if (range.location != NSNotFound)
                      {
                          filename = [disposition substringFromIndex:range.location+9];
                          range = [filename rangeOfString:@".o8d" options:NSCaseInsensitiveSearch];
                          if (range.location != NSNotFound)
                          {
                              filename = [filename substringToIndex:range.location];
                              filename = [filename stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLPathAllowedCharacterSet];
                          }
                      }
                      
                      ok = [self parseOctgnDeck:responseObject name:filename];
                  }
                  [self downloadFinished:ok];
              }
              failure:^(AFHTTPRequestOperation* operation, NSError* error) {
                  @strongify(self);
                  // NSLog(@"download failed %@", operation);
                  [self downloadFinished:NO];
              }
     ];
}

-(void) downloadFinished:(BOOL)ok
{
    [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
    [self.alert dismissWithClickedButtonIndex:-1 animated:NO];
}

-(BOOL) parseJsonDecklist:(NSDictionary*) decklist
{
    Deck* deck = [[Deck alloc] init];
    
    deck.name = [decklist objectForKey:@"name"];
    
    NSString* notes = [decklist objectForKey:@"description"];
    if (notes.length > 0)
    {
        notes = [notes stringByReplacingOccurrencesOfString:@"<p>" withString:@""];
        notes = [notes stringByReplacingOccurrencesOfString:@"</p>" withString:@""];
        deck.notes = notes;
    }
    
    NSDictionary* cards = [decklist objectForKey:@"cards"];
    for (NSString* code in [cards allKeys])
    {
        int qty = [[cards objectForKey:code] intValue];
        Card* card = [CardManager cardByCode:code];
        if (card)
        {
            if (card.type == NRCardTypeIdentity)
            {
                deck.role = card.role;
            }
            [deck addCard:card copies:qty];
        }
    }
    
    if (deck.identity != nil && deck.cards.count > 0)
    {
        [[NSNotificationCenter defaultCenter] postNotificationName:Notifications.IMPORT_DECK object:self userInfo:@{ @"deck": deck }];
        return YES;
    }
    return NO;
}

-(BOOL) parseOctgnDeck:(NSXMLParser*)parser name:(NSString*)name
{
    OctgnImport* importer = [[OctgnImport alloc] init];
    Deck* deck = [importer parseOctgnDeckWithParser:parser];
    
    if (deck.identity != nil && deck.cards.count > 0)
    {
        // NSLog(@"got deck: %@ %d", deck.identity.name, deck.cards.count);
        // NSLog(@"name: %@", filename);
        deck.name = name;
        [[NSNotificationCenter defaultCenter] postNotificationName:Notifications.IMPORT_DECK object:self userInfo:@{ @"deck": deck }];
        return YES;
    }
    return NO;
}

+(void) importDeckFromLocalUrl:(NSURL*)url
{
    NSString* b64 = [url.path substringFromIndex:1]; // strip off the leading "/" character
    NSData* data = [[NSData alloc] initWithBase64EncodedString:b64 options:0];
    NSData* uncompressed = [GZip gzipInflate:data];
    NSString* deckStr = [[NSString alloc] initWithData:uncompressed encoding:NSUTF8StringEncoding];
    
    NSArray* parts = [deckStr componentsSeparatedByString:@"&"];
    Deck* deck = [[Deck alloc] init];
    for (NSString* card in parts)
    {
        NSArray* cq = [card componentsSeparatedByString:@"="];
        NSString* code = cq[0];
        NSString* qty = cq[1];
        
        if ([code isEqualToString:@"name"])
        {
            deck.name = [qty stringByRemovingPercentEncoding];            
        }
        else
        {
            Card* card = [CardManager cardByCode:code];
            if (card && qty.integerValue > 0)
            {
                [deck addCard:card copies:qty.integerValue];
            }
        }
    }
    
    [[NSNotificationCenter defaultCenter] postNotificationName:Notifications.IMPORT_DECK object:self userInfo:@{ @"deck": deck }];
}

@end

File: NetDeck/DeckImport.swift
//
//  DeckImport.swift
//  NetDeck
//
//  Created by Gereon Steffens on 21.02.16.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import UIKit
import SDCAlertView
import Alamofire
import SwiftyUserDefaults

final class DeckImport: NSObject {
    
    private enum DeckBuilderSource {
        case none
        case nrdbList
        case nrdbShared
        case meteor
    }

    private struct DeckSource {
        let deckId: String
        let source: DeckBuilderSource
        let hash: String?

        init(deckId: String, source: DeckBuilderSource, hash: String? = nil) {
            self.deckId = deckId
            self.source = source
            self.hash = hash
        }
    }
    
    static let sharedInstance = DeckImport()
    
    private let importAlways = false // set to true for easier debugging
    
    private var deck: Deck?
    private var deckSource: DeckSource?
    private var uiAlert: UIAlertController?
    private var sdcAlert: AlertController?
    private var downloadStopped: Bool!
    private var request: Request?
 
    static func updateCount() {
        Defaults[.clipChangeCount] = UIPasteboard.general.changeCount
    }
    
    static func checkClipboardForDeck() {
        return sharedInstance.checkClipboardForDeck()
    }
    
    func checkClipboardForDeck() {
        let alwaysImport = BuildConfig.debug ? importAlways : false
        
        let pasteboard = UIPasteboard.general
        let lastChange = Defaults[.clipChangeCount]
        if lastChange == pasteboard.changeCount && !alwaysImport {
            return
        }
        Defaults[.clipChangeCount] = pasteboard.changeCount
        
        guard let clip = pasteboard.string else { return }
        if clip.count == 0 {
            return
        }

        let lines = clip.components(separatedBy: CharacterSet.newlines)
        
        self.deck = nil
        
        self.deckSource = self.checkForNetrunnerDbDeckURL(lines)
        if self.deckSource == nil {
            self.deckSource = self.checkForMeteorDeckURL(lines)
        }
        
        self.uiAlert = nil
        if let deckSource = self.deckSource {
            switch deckSource.source {
            case .nrdbList, .nrdbShared:
                let msg = "Detected a NetrunnerDB.com deck list URL in your clipboard. Download and import this deck?".localized()
                self.uiAlert = UIAlertController(title: nil, message: msg, preferredStyle: .alert)
                
            case .meteor:
                let msg = "Detected a Meteor deck list URL in your clipboard. Download and import this deck?".localized()
                self.uiAlert = UIAlertController(title: nil, message: msg, preferredStyle: .alert)
                
            default:
                break
            }
        } else {
            self.deck = self.checkForTextDeck(lines)
            
            if self.deck != nil {
                let msg = "Detected a deck list in your clipboard. Import this deck?".localized()
                self.uiAlert = UIAlertController(title: nil, message: msg, preferredStyle: .alert)
            }
        }
        
        if let alert = self.uiAlert {
            alert.addAction(UIAlertAction(title:"No".localized(), style: .cancel, handler: nil))
            alert.addAction(UIAlertAction(title:"Yes".localized(), style: .default) { (UIAlertAction) -> Void in
                Analytics.logEvent(.importFromClipboard)
                if self.deck != nil {
                    NotificationCenter.default.post(name: Notifications.importDeck, object:self, userInfo:["deck": self.deck!])
                }
                else if self.deckSource != nil {
                    self.downloadDeck(self.deckSource!)
                }
                self.deck = nil
                self.deckSource = nil
            })
            
            alert.show()
        }
    }
    
    private func checkForNetrunnerDbDeckURL(_ lines: [String]) -> DeckSource? {
        // a netrunnerdb.com decklist url looks like this:
        // https://netrunnerdb.com/en/decklist/3124/in-a-red-dress-and-alone-jamieson-s-store-champ-deck-#
        // or like this:
        // https://netrunnerdb.com/en/deck/view/456867
        // or like this (added hash after the scraping exploit)
        // https://netrunnerdb.com/en/deck/view/456867/hash
        
        let list = try! NSRegularExpression(pattern: "https://netrunnerdb.com/../decklist/(\\d*)/.*", options:[])
        // let shared1 = try! NSRegularExpression(pattern: "https://netrunnerdb.com/../deck/view/(\\d*)/(.*)", options:[])
        let shared2 = try! NSRegularExpression(pattern: "https://netrunnerdb.com/../deck/view/(\\d*)", options:[])
        // FIXME("wip until hashed url format is final")

        let dict = [
            DeckBuilderSource.nrdbShared: [ shared2 ],
            DeckBuilderSource.nrdbList: [ list ]
        ]

        for entry in dict {
            for regex in entry.value {
                for line in lines {
                    if let match = regex.firstMatch(in: line, options:[], range: NSMakeRange(0, line.count)), match.numberOfRanges > 1 {
                        let l = line as NSString
                        let hash = match.numberOfRanges == 3 ? l.substring(with: match.range(at: 2)) : nil
                        let src = DeckSource(deckId: l.substring(with: match.range(at: 1)), source: entry.key, hash: hash)
                        return src
                    }
                }
            }
        }

        return nil
    }
    
    private func checkForMeteorDeckURL(_ lines: [String]) -> DeckSource? {
        // a meteor.stimhack.com decklist url looks like this:
        // https://meteor.stimhack.com/decks/yBMJ3GL6FPozt9nkQ/
        // or like this (no slash)
        // https://meteor.stimhack.com/decks/i6sLkn5cYZ3633WAu
        
        let regEx = try! NSRegularExpression(pattern:"https://meteor.stimhack.com/decks/(.*)/?", options:[])
        
        for line in lines {
            if let match = regEx.firstMatch(in: line, options:[], range:NSMakeRange(0, line.count)) , match.numberOfRanges == 2 {
                let l = line as NSString
                let src = DeckSource(deckId: l.substring(with: match.range(at: 1)), source: .meteor)
                return src
            }
        }
        
        return nil
    }

    private func findMatch(in line: String, regexes: [NSRegularExpression]) -> NSTextCheckingResult? {
        for regex in regexes {
            if let match = regex.firstMatch(in: line, options: [], range: NSMakeRange(0, line.count)) {
                return match
            }
        }
        return nil
    }
    
    private func checkForTextDeck(_ lines: [String]) -> Deck? {
        let cards = CardManager.allCards()

        let regex1 = try! NSRegularExpression(pattern:"^([0-9])x", options:[]) // start with "1x ..."
        let regex2 = try! NSRegularExpression(pattern:" x([0-9])", options:[]) // end with "... x3"
        let regex3 = try! NSRegularExpression(pattern:"^([0-9]) ", options:[]) // start with "1 ..."
        let regexes = [regex1, regex2, regex3]
        
        var name: String?
        let deck = Deck(role: .none)
        for line in lines {
            if name == nil {
                name = line
            }
            
            for var c in cards {
                // don't bother checking cards of the opposite role (as soon as we know this deck's role)
                let roleOk = deck.role == .none || deck.role == c.role
                if !roleOk {
                    continue
                }
                
                let range =
                    line.range(of: c.englishName, options:[.caseInsensitive,.diacriticInsensitive]) ??
                    line.range(of: c.name, options:[.caseInsensitive,.diacriticInsensitive])
                
                if range != nil {
                    if c.type == .identity {
                        deck.addCard(c, copies:1)
                        // NSLog(@"found identity %@", c.name);
                    } else {
                        if let match = self.findMatch(in: line, regexes: regexes), match.numberOfRanges == 2 {
                            let l = line as NSString
                            let count = l.substring(with: match.range(at: 1))
                            // NSLog(@"found card %@ x %@", count, c.name);

                            if Defaults[.useCore2], let newCode = Card.originalToRevised[c.code] {
                                c = CardManager.cardBy(newCode) ?? c
                            }
                            
                            let max = deck.isDraft ? 100 : 4;
                            if let cnt = Int(count), cnt > 0 && cnt < max {
                                deck.addCard(c, copies: cnt)
                            }
                            
                            break
                        }
                    }
                }
            }
        }
        
        if deck.identity != nil && deck.cards.count > 0 {
            deck.name = name ?? ""
            return deck
        } else {
            return nil
        }
    }
    
    private func downloadDeck(_ source: DeckSource) {
        let alert = AlertController(title: "Downloading Deck".localized(), message: nil, preferredStyle: .alert)
        alert.visualStyle = CustomAlertVisualStyle(alertStyle: .alert)
        self.sdcAlert = alert

        var style = UIActivityIndicatorView.Style.gray
        if #available(iOS 13, *) {
            style = .medium
        }
        let spinner = UIActivityIndicatorView(style: style)
        spinner.translatesAutoresizingMaskIntoConstraints = false
        spinner.startAnimating()
        
        alert.contentView.addSubview(spinner)
        
        spinner.centerXAnchor.constraint(equalTo: alert.contentView.centerXAnchor).isActive = true
        spinner.topAnchor.constraint(equalTo: alert.contentView.topAnchor).isActive = true
        spinner.bottomAnchor.constraint(equalTo: alert.contentView.bottomAnchor).isActive = true
        
        alert.addAction(AlertAction(title: "Stop".localized(), style: .normal, handler: { action in
            self.downloadStopped = true
            self.sdcAlert = nil
            if let req = self.request {
                req.cancel()
            }
        }))
        
        alert.present()

        DispatchQueue.main.async {
            switch source.source {
            case .nrdbList:
                self.doDownloadDeckFromNetrunnerDbList(source)

            case .nrdbShared:
                self.doDownloadDeckFromNetrunnerDbShared(source)

            case .meteor:
                self.doDownloadDeckFromMeteor(source)

            case .none:
                break
            }
        }
    }
    
    private func doDownloadDeckFromNetrunnerDbList(_ source: DeckSource) {
        let deckUrl = "https://netrunnerdb.com/api/2.0/public/decklist/" + source.deckId
        self.doDownloadDeckFromNetrunnerDb(deckUrl)
    
    }

    private func doDownloadDeckFromNetrunnerDbShared(_ source: DeckSource) {
        let deckUrl = "https://netrunnerdb.com/api/2.0/public/deck/" + source.deckId
//        if let hash = source.hash {
//            deckUrl += "/" + hash
//        }
        self.doDownloadDeckFromNetrunnerDb(deckUrl)
    }
    
    func doDownloadDeckFromNetrunnerDb(_ deckUrl: String) {
        self.downloadStopped = false
        
        self.request = Alamofire.request(deckUrl).validate().responseJSON { response in
            switch response.result {
            case .success:
                if let data = response.data, !self.downloadStopped {
                    let decks = Deck.arrayFromJson(data)
                    self.downloadFinished(decks.first)
                }
            case .failure(let error):
                print(error)
                self.downloadFinished(nil)
            }
        }
    }
    
    private func doDownloadDeckFromMeteor(_ source: DeckSource) {
        let deckUrl = "https://meteor.stimhack.com/deckexport/octgn/" + source.deckId
        self.downloadStopped = false
        
        self.request = Alamofire.request(deckUrl).responseData(completionHandler: { (response) in
            switch response.result {
            case .success:
                if let value = response.result.value {
                    var filename: NSString = ""
                    if let disposition = response.response?.allHeaderFields["Content-Disposition"] as? NSString {
                        var range = disposition.range(of: "filename=", options: .caseInsensitive)
                        if range.location != NSNotFound {
                            filename = disposition.substring(from: range.location+9) as NSString
                            range = filename.range(of: ".o8d")
                            if range.location != NSNotFound {
                                filename = filename.substring(to: range.location) as NSString
                            }
                            filename = filename.removingPercentEncoding! as NSString
                        }
                    }
                    let deck = self.importOctgnDeck(value, name: filename as String)
                    
                    self.downloadFinished(deck)
                }
            case .failure(let error):
                print(error)
                self.downloadFinished(nil)
            }
        })
        
     }

    func importOctgnDeck(_ data: Data, name: String) -> Deck? {
        let importer = OctgnImport()
        if let deck = importer.parseOctgnDeckFromData(data) {
            if deck.identity != nil && deck.cards.count > 0 {
                deck.name = name
                return deck
            }
        }
        return nil
    }

    private func downloadFinished(_ deck: Deck?) {
        self.request = nil
        if let alert = self.sdcAlert {
            alert.dismiss(animated: true) {
                // send notification when dismissal is done
                self.postImportNotification(deck)
            }
        } else {
            self.postImportNotification(deck)
        }
    }

    private func postImportNotification(_ deck: Deck?) {
        if let deck = deck {
            NotificationCenter.default.post(name: Notifications.importDeck, object: self, userInfo: ["deck": deck])
        }
    }
}
